### PR TITLE
Simplified how Metadata is Stored on Types Which Support Forward Declaration

### LINKED ISSUE #3039: Handling of deprecated in Slice library
The new Slice/deprecated test (see #3038) shows two issues in the Slice library (aka Slice parser).

- `getDeprecationReason` returns an empty string and not `nullopt` for Slice metadata `["deprecated"]`
-  when we define a Slice class or interface (Def) without an existing Decl, the deprecated metadata for the Def is not copied into the Decl created by the Parser

I think these two issues are related. It could be cleaner for the Parser to consume the deprecated Metadata and provide direct support for deprecated--and not keep it as a generic metadata to be consumed by the Slice compilers.

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
The main focus of this PR is #3216 (cleaning up how metadata is stored)
but it also fixes #3039 (which was specifically about how `deprecated` is handled between `Decl` and `Def`).

----

1) **Added a new `appendMetadata` function.**
So now we have `setMetadata` which overwrites existing metadata with a new list (_used by the validator where we want to throw away bad metadata_) and `appendMetadata` which just appends metadata to a Slice entity without touching any that's already present (_used by the parser to add metadata to entities_).
This fixes the problem where forward-declarations were constantly overwriting their own metadata.

2) **All class/interface metadata is now stored exclusively on the `Decl` types, and never the `Def` types.**
Previously, both of them had their own distinct lists of metadata. Note that the `Def` types still support the same API for metadata, but these functions just forward to the 'underlying' `Decl` type. This way the code-gen never has to worry about whether it should be using `Decl` or `Def`, both will work just fine.

I also removed all of the `visitClassDef`/`visitInterfaceDef` metadata validation functions.
It is sufficient to only visit the `Decl` types now.

### CODE REVIEW

File: cpp/src/Slice/MetadataValidation.cpp [hunk 1]
```diff
@@ -95,26 +94,27 @@ Slice::validateMetadata(const UnitPtr& p, string_view prefix, map<string, Metada
          typeid(DataMember)},
         MetadataArgumentKind::OptionalTextArgument,
     };
+    knownMetadata.emplace("deprecate", std::move(deprecatedInfo)); // Kept as an alias for 'deprecated'.
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/MetadataValidation.cpp:97]: I forgot to add this in my original metadata validation PR.
We don't actually use it in our Slice files, but we decided to keep support for the `3.7` name for backwards compatibility. We don't want to bother our users about adding a single `d` character to their metadata.

File: cpp/src/Slice/MetadataValidation.cpp [hunk 2]
```diff
@@ -73,17 +74,15 @@ Slice::validateMetadata(const UnitPtr& p, string_view prefix, map<string, Metada
 
     // "amd"
     MetadataInfo amdInfo = {
-        {typeid(InterfaceDef), typeid(Operation)},
+        {typeid(InterfaceDecl), typeid(Operation)},
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/MetadataValidation.cpp:77]: Now that all metadata is on the `Decl` types, these are the types that should be passed into the validator.
I added a note in the header file mentioning this fact.

File: cpp/src/Slice/Parser.cpp [hunk 3]
```diff
@@ -1023,16 +1036,17 @@ Slice::Contained::isDeprecated() const
 optional<string>
 Slice::Contained::getDeprecationReason() const
 {
+    string reasonMessage;
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:1039]: This is unrelated to the main goal of this PR, but is small, and fixes #3039.

If no message was supplied to `deprecated` metadata, this should return `nullopt`, not `""`.
And now it does!

File: cpp/src/Slice/Parser.cpp [hunk 4]
```diff
@@ -1178,23 +1192,15 @@ Slice::Container::createClassDef(const string& name, int id, const ClassDefPtr&
     }
 
     ClassDefPtr def = make_shared<ClassDef>(shared_from_this(), name, id, base);
-    _unit->addContent(def);
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:1181]: Whenever you define a class, we always make sure there's a corresponding `Decl`. If there isn't, we create one here.

Previously, we would add the `Def` to the module first, then add the `Decl` 2nd.
With my changes though, this causes metadata warnings to be emitted in a lame order.

I swapped the order, so now we add the `Decl` first, and then add the `Def` afterwards.
This fixes the order, and also... it's just simpler.

File: cpp/src/Slice/Parser.cpp [hunk 5]
```diff
@@ -1330,23 +1336,15 @@ Slice::Container::createInterfaceDef(const string& name, const InterfaceList& ba
     InterfaceDecl::checkBasesAreLegal(name, bases, _unit);
 
     InterfaceDefPtr def = make_shared<InterfaceDef>(shared_from_this(), name, bases);
-    _unit->addContent(def);
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:1333]: Same for interface. Now we add the `Decl` before we add the `Def`.
This is more expected from the POV of the visitors.

File: cpp/src/Slice/Parser.cpp [hunk 6]
```diff
@@ -2695,6 +2693,24 @@ Slice::ClassDef::compactId() const
     return _compactId;
 }
 
+MetadataList
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:2696]: The metadata API on the `Def` types just always forward to their corresponding `Decl`.

File: cpp/src/Slice/Parser.cpp [hunk 7]
```diff
@@ -4713,10 +4747,7 @@ Slice::Unit::addFileMetadata(MetadataList metadata)
     }
     else
     {
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:4749]: Might as well use the new `appendMetadata` API.

File: cpp/src/Slice/Parser.cpp [hunk 8]
```diff
@@ -214,10 +214,17 @@ Slice::DefinitionContext::setMetadata(MetadataList metadata)
     initSuppressedWarnings();
 }
 
+void
+Slice::DefinitionContext::appendMetadata(MetadataList metadata)
+{
+    _metadata.splice(_metadata.end(), metadata);
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:220]: missing std::move

File: cpp/src/Slice/Parser.cpp [hunk 9]
```diff
@@ -965,10 +972,16 @@ Slice::Contained::setMetadata(MetadataList metadata)
     _metadata = metadata;
 }
 
+void
+Slice::Contained::appendMetadata(MetadataList metadata)
+{
+    _metadata.splice(_metadata.end(), metadata);
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:978]: missing std::move

File: cpp/src/Slice/Parser.cpp [hunk 10]
```diff
@@ -4713,10 +4747,7 @@ Slice::Unit::addFileMetadata(MetadataList metadata)
     }
     else
     {
-        // Append the file metadata to any existing metadata (e.g., default file metadata).
-        MetadataList l = dc->getMetadata();
-        move(metadata.begin(), metadata.end(), back_inserter(l));
-        dc->setMetadata(l);
+        dc->appendMetadata(metadata);
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:4750]: std::move

**Review by bernardnormier — APPROVED** [review]

**Review by externl — APPROVED** [review]

File: cpp/src/Slice/Parser.cpp (hunk 9 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:978]: Fixed.

File: cpp/src/Slice/Parser.cpp (hunk 10 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:4750]: Fixed.

**Pull request merged**